scilla/src/call_scilla_server.rs

```diff
@@ -0,0 +1,275 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub result: Option<Value>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub error: Option<JsonRpcError>,
+    id: u32,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct JsonRpcError {
+    code: i32,
+    message: String,
+    data: Option<Value>,
+}
+
+fn respond_json(val: Value, mut connection: &TcpStream, id: u32) {
+    let response = JsonRpcResponse {
+        jsonrpc: "2.0".to_string(),
+        result: Some(val),
+        error: None,
+        id,
+    };
+
+    let response_str = serde_json::to_string(&response).unwrap();
+    let response_str = response_str.to_owned() + "\n";
+
+    connection.write_all(response_str.as_bytes()).unwrap();
+}
+
+pub fn call_scilla_server<B: evm::backend::Backend>(
+    method: &str,
+    params: Params,
+    tcp_scilla_server: &mut ScillaServer<B>,
+) -> Result<JsonRpcResponse> {
+    let request = JsonRpcRequest::new(method, params, 1);
+    let request_str = serde_json::to_string(&request)?;
+    let request_str = request_str + "\n"; // This is required to complete the request
+
+    trace!("Connecting to scilla server...");
+    let mut stream = TcpStream::connect("127.0.0.1:12345")?;
+    let mut stream_backend = TcpStream::connect("127.0.0.1:12346")?;
+
+    trace!("Sending request to scilla server: {:?}", request_str);
+    stream.write_all(request_str.as_bytes())?;
+
+    // Set non-blocking so that we can check and respond to both connections (and timeout).
+    stream.set_nonblocking(true)?;
+    stream_backend.set_nonblocking(true)?;
+
+    let mut response = vec![];
+    let mut response_backend = vec![];
+
+    let mut bytes_read = 0;
+    let mut bytes_read_backend = 0;
+
+    let now = std::time::Instant::now();
+
+    loop {
+        if now.elapsed().as_millis() > 5000 {
+            return Err(anyhow!("Timeout waiting for response from scilla server"));
+        }
+
+        // Bump up the buffers we are reading into if we are at the end (exponential growth)
+        if bytes_read >= response.len() {
+            response.resize((response.len() + 1) * 2, 0);
+        }
+
+        if bytes_read_backend >= response_backend.len() {
+            response_backend.resize((response_backend.len() + 1) * 2, 0);
+        }
+
+        // Read backend request
+        let bytes_read_ch2 = stream_backend.read(&mut response_backend[bytes_read_backend..]);
+
+        match bytes_read_ch2 {
+            Ok(bytes_read) => {
+                bytes_read_backend += bytes_read;
+            }
+            // It is valid if there is no data to read for either connection because it can still be
+            // processing the request.
+            Err(_) => {
+                //debug!("Scilla backend error: {:?}", e);
+            }
+        }
+
+        // Check if final character read was a newline, which indicates the end of the response
+        if bytes_read_backend > 0 && response_backend[bytes_read_backend - 1] == b'\n' {
```

Comment:
This is all quite scary and manual to me, but maybe we can live with it until we reach "Scilla interface optimisation".. :)

Reply:
Yes, I agree with you on that. It can definetly be improved, but for the moment most of the development work is not on the interface itself (but rather feature completeness).

I realised later I should have upgraded the scilla connection to http rather than tcp, then the jsonrpc framework would support it.